src/qiskit_qasm3_import/expression.py

```diff
@@ -17,19 +17,37 @@
 
 __all__ = ["ValueResolver", "resolve_condition"]
 
-from typing import Any, Iterable, Mapping, Tuple, Union
+import re
+from typing import Any, Iterable, Tuple, Union
 
 from openqasm3 import ast
 from openqasm3.visitor import QASMVisitor
-from qiskit.circuit import Clbit
+from qiskit.circuit import Clbit, Qubit
 
 from . import types
-from .exceptions import raise_from_node
-from .data import Symbol
+from .exceptions import raise_from_node, PhysicalQubitInGateError
+from .data import Symbol, Scope
+from .state import State, SymbolTable
 
 
 _IntegerT = Union[types.Never, types.Int, types.Uint]
 
+_PHYSICAL_QUBIT_RE = re.compile(r"\$\d+")
+
+
+def is_physical(name: Union[str, Symbol]):
+    "Return true if name is a valid identifier for a physical qubit."
+    if isinstance(name, Symbol):
+        name = name.name
+    return re.match(_PHYSICAL_QUBIT_RE, name) is not None
+
+
+# A hardware-qubit symbol has the form '$' followed by digits.
+# The digits are the identifier used by backends.
+def hardware_qubit_map(symbol_table: SymbolTable):
+    "Return a `dict` mapping `Qubit` instances to `int`s representing physical qubit identifiers."
+    return {sym.data: int(sym.name[1:]) for sym in symbol_table.globals() if is_physical(sym)}
```

Comment:
Probably nit-picky, but since this is only used in `converter.py`, I think it logically belongs there?

I do still think that the number of hardware qubits is important for useful reconstruction.  We might be able to make a valid `Layout` object with holes in it, but in practice, that doesn't fully represent a hardware-mapped circuit, because those `QuantumCircuit` objects always have as many qubits in them as the backend has, where the unused qubits are virtually mapped to a (non-existent) quantum register called `ancilla`:
```python
In [1]: from qiskit import QuantumCircuit, transpile
   ...: from qiskit.providers.fake_provider import FakeKolkata
   ...:
   ...: qc = transpile(QuantumCircuit(2), FakeKolkata())
   ...: print(qc.num_qubits)
   ...: print(qc._layout.initial_layout)
27
Layout({
0: Qubit(QuantumRegister(2, 'q'), 0),
1: Qubit(QuantumRegister(2, 'q'), 1),
2: Qubit(QuantumRegister(25, 'ancilla'), 0),
3: Qubit(QuantumRegister(25, 'ancilla'), 1),
4: Qubit(QuantumRegister(25, 'ancilla'), 2),
5: Qubit(QuantumRegister(25, 'ancilla'), 3),
6: Qubit(QuantumRegister(25, 'ancilla'), 4),
7: Qubit(QuantumRegister(25, 'ancilla'), 5),
8: Qubit(QuantumRegister(25, 'ancilla'), 6),
9: Qubit(QuantumRegister(25, 'ancilla'), 7),
10: Qubit(QuantumRegister(25, 'ancilla'), 8),
11: Qubit(QuantumRegister(25, 'ancilla'), 9),
12: Qubit(QuantumRegister(25, 'ancilla'), 10),
13: Qubit(QuantumRegister(25, 'ancilla'), 11),
14: Qubit(QuantumRegister(25, 'ancilla'), 12),
15: Qubit(QuantumRegister(25, 'ancilla'), 13),
16: Qubit(QuantumRegister(25, 'ancilla'), 14),
17: Qubit(QuantumRegister(25, 'ancilla'), 15),
18: Qubit(QuantumRegister(25, 'ancilla'), 16),
19: Qubit(QuantumRegister(25, 'ancilla'), 17),
20: Qubit(QuantumRegister(25, 'ancilla'), 18),
21: Qubit(QuantumRegister(25, 'ancilla'), 19),
22: Qubit(QuantumRegister(25, 'ancilla'), 20),
23: Qubit(QuantumRegister(25, 'ancilla'), 21),
24: Qubit(QuantumRegister(25, 'ancilla'), 22),
25: Qubit(QuantumRegister(25, 'ancilla'), 23),
26: Qubit(QuantumRegister(25, 'ancilla'), 24)
})
```

I view this support of hardware qubits as moving us towards being able to replace the direct-string OQ3 input form for `dynamic=True` backends in IBM provider, which is currently passed directly to the QSS stack.  To be as close to a drop-in replacement for that as possible, we'll need to make it easy to produce a `QuantumCircuit` object that could be the direct output of `transpile`, which in this case means being full-width on the backend.

Reply:
> Probably nit-picky, but since this is only used in converter.py, I think it logically belongs there?

I'm fine with moving it to converter.py. This will require importing `SymbolTable` and `is_physical` into converter.py

Reply:
Fixed in 9199a6a . Also, after removing spurious `assert`s, I am able to remove `SymbolTable` from imports into expression.py

Reply:
Fixed `_STDGATES` location in bf80706